File: vending_machine/python/vending_machine.py
class Item:
    def __init__(self, item_id: int, name: str, price: int):
        self.item_id = item_id
        self.name = name
        self.price = price

class Slot:
    def __init__(self, slot_id: str, item: Item, quantity: int):
        self.slot_id = slot_id
        self.item = item
        self.quantity = quantity

class VendingMachine:
    def __init__(self):
        self.slots: dict[str, Slot] = {}
        self.inserted_balance: int = 0
    
    def add_slot(self, slot: Slot) -> None:
        if slot.slot_id in self.slots:
            raise ValueError(f"slot {slot.slot_id} already exists")
        self.slots[slot.slot_id] = slot
    
    def load_slots(self, slots: list[Slot]) -> None:
        for slot in slots:
            self.add_slot(slot)
    
    def display_slots(self) -> None:
        for slot_id, slot in self.slots.items():
            item = slot.item
            print(
                f"slot ID: {slot_id}, "
                f"item name: {item.name}, "
                f"item price: ${item.price / 100:.2f}, "
                f"quantity: {slot.quantity}"
            )
    
    def insert_money(self, amount: int) -> None:
        if amount <= 0:
            raise ValueError("amount must be positive")
        self.inserted_balance += amount
    
    def return_change(self) -> int:
        change = self.inserted_balance
        self.inserted_balance = 0
        return change

    def purchase_item(self, slot_id: str) -> tuple[Item | None, int]:
        if slot_id not in self.slots:
            print("invalid slot id")
            change = self.return_change()
            return None, change

        slot = self.slots[slot_id]

        if slot.quantity == 0:
            print("selected item out of stock")
            change = self.return_change()
            return None, change

        item = slot.item

        if item.price > self.inserted_balance:
            print("insufficient balance")
            return None, 0

        self.inserted_balance -= item.price
        slot.quantity -= 1

        change = self.return_change()
        return item, change

"""
In Java and C++, "main" has a special meaning 
so the runtime will specifically look for main
as the entry point.

In Python, main is not special and is just a 
normal function name.

So we must call main as our entry point, which is 
why we have something like __name__ == "__main__"... .

What is __name__?
Each Python file is treated as a module, and Python
gives each module a built-in variable called __name__.

__name__ is set to be "__main__" if the file runs directly,
or to the module's actual name if the file is imported.
"""

def initialize_vending_machine():
    vm = VendingMachine()

    coke = Item(item_id=0, name="coke", price=150)
    sprite = Item(item_id=1, name="sprite", price=200)
    water = Item(item_id=2, name="water", price=100)
    iced_tea = Item(item_id=3, name="iced tea", price=175)
    lemonade = Item(item_id=4, name="lemonade", price=175)
    orange_juice = Item(item_id=5, name="orange juice", price=225)
    coffee = Item(item_id=6, name="coffee", price=250)
    energy_drink = Item(item_id=7, name="energy drink", price=300)
    chips = Item(item_id=8, name="chips", price=125)
    candy_bar = Item(item_id=9, name="candy bar", price=150)
    cookies = Item(item_id=10, name="cookies", price=175)
    gum = Item(item_id=11, name="gum", price=75)

    items = [
        coke,
        sprite,
        water,
        iced_tea,
        lemonade,
        orange_juice,
        coffee,
        energy_drink,
        chips,
        candy_bar,
        cookies,
        gum,
    ]

    slots = []
    rows = ["A", "B", "C", "D", "E", "F", "G", "H"]
    cols = [1, 2, 3, 4, 5, 6]

    item_index = 0

    for row in rows:
        for col in cols:
            slot_id = f"{row}{col}"
            item = items[item_index % len(items)]
            quantity = 5

            slots.append(Slot(slot_id=slot_id, item=item, quantity=quantity))

            item_index += 1

    vm.load_slots(slots)

    return vm

def main():
    vm = initialize_vending_machine()
    vm.display_slots()

    amount = 300
    vm.insert_money(amount)
    print(f"Inserted ${amount / 100:.2f}.")

    item, change = vm.purchase_item("A1")
    print(f"Purchased {item.name}, received ${change / 100:.2f} as change.")

    assert item is not None
    assert item.name == "coke"
    assert change == 150
    assert vm.slots["A1"].quantity == 4
    assert vm.inserted_balance == 0

if __name__ == "__main__":
    main()

File: vending_machine/cpp/vending_machine.cpp
#include <iostream>      // For console input/output, mainly std::cout and std::endl.
#include <string>        // For std::string.
#include <vector>        // For std::vector, a dynamic array.
#include <unordered_map> // For std::unordered_map, a hash map / dictionary.
#include <optional>      // For std::optional, used to represent a value that may or may not exist.
#include <utility>       // For std::pair, used to return two values together.
#include <stdexcept>     // For standard exceptions like std::invalid_argument.
#include <iomanip>       // For output formatting, such as std::fixed and std::setprecision.
#include <cassert>       // For assert(), used in simple tests/debug checks.

// compile using C++17: c++ -std=c++17 vending_machine.cpp -o vending_machine

class Item {
public:
    int item_id;
    std::string name;
    int price;

    /*
        Item(int item_id, std::string name, int price) {
            this->item_id = item_id;
            this->name = name;
            this->price = price;
        }
    */

    // this is a more idiomatic constructor syntax
    Item(int item_id, std::string name, int price)
        : item_id(item_id), name(name), price(price) {
            // do something here
        }
};

class Slot {
public:
    std::string slot_id;
    Item item;
    int quantity;

    Slot(std::string slot_id, Item item, int quantity)
        : slot_id(slot_id), item(item), quantity(quantity) {}
};

class VendingMachine {
private:
    std::unordered_map<std::string, Slot> slots;
    int inserted_balance;

public:
    VendingMachine()
        : inserted_balance(0) {}
    
    void add_slot(const Slot& slot) {
        if (slots.find(slot.slot_id) != slots.end()) {
            throw std::invalid_argument("slot " + slot.slot_id + " already exists");
        }
        slots.emplace(slot.slot_id, slot);
    }

    void load_slots(const std::vector<Slot>& new_slots) {
        for (const Slot& slot : new_slots) {
            add_slot(slot);
        }
    }

    void display_slots() const {
        for (const auto& pair : slots) {
            const std::string& slot_id = pair.first;
            const Slot& slot = pair.second;
            const Item& item = slot.item;

            std::cout 
                << "slot ID: " << slot_id
                << ", item name: " << item.name
                << ", item price: $" << std::fixed << std::setprecision(2)
                << item.price / 100.0
                << ", quantity: " << slot.quantity
                << std::endl;
        }
    }

    void insert_money(int amount) {
        if (amount <= 0) {
            throw std::invalid_argument("amount must be positive");
        }
        inserted_balance += amount;
    }

    int return_change() {
        int change = inserted_balance;
        inserted_balance = 0;
        return change;
    }

    std::pair<std::optional<Item>, int> purchase_item(const std::string& slot_id) {
        if (slots.find(slot_id) == slots.end()) {
            std::cout << "invalid slot id" << std::endl;
            int change = return_change();
            return {std::nullopt, change};
        }

        Slot& slot = slots.at(slot_id);

        if (slot.quantity == 0) {
            std::cout << "selected item out of stock " << std::endl;
            int change = return_change();
            return {std::nullopt, change};
        }

        Item item = slot.item;

        if (item.price > inserted_balance) {
            std::cout << "insufficient balance" << std::endl;
            return {std::nullopt, 0};
        }

        inserted_balance -= item.price;
        slot.quantity -= 1;

        int change = return_change();
        return {item, change};
    }

    int get_inserted_balance () const {
        return inserted_balance;
    }

    int get_slot_quantity(const std::string& slot_id) const {
        return slots.at(slot_id).quantity;
    }
};

// Testing

VendingMachine initialize_vending_machine() {
    VendingMachine vm;

    Item coke(0, "coke", 150);
    Item sprite(1, "sprite", 200);
    Item water(2, "water", 100);
    Item iced_tea(3, "iced tea", 175);
    Item lemonade(4, "lemonade", 175);
    Item orange_juice(5, "orange juice", 225);
    Item coffee(6, "coffee", 250);
    Item energy_drink(7, "energy drink", 300);
    Item chips(8, "chips", 125);
    Item candy_bar(9, "candy bar", 150);
    Item cookies(10, "cookies", 175);
    Item gum(11, "gum", 75);

    std::vector<Item> items = {
        coke,
        sprite,
        water,
        iced_tea,
        lemonade,
        orange_juice,
        coffee,
        energy_drink,
        chips,
        candy_bar,
        cookies,
        gum
    };

    std::vector<Slot> slots;
    std::vector<char> rows = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'};
    std::vector<int> cols = {1, 2, 3, 4, 5, 6};

    int item_index = 0;

    for (char row : rows) {
        for (int col : cols) {
            std::string slot_id = std::string(1, row) + std::to_string(col);
            Item item = items[item_index % items.size()];
            int quantity = 5;

            slots.emplace_back(slot_id, item, quantity);

            item_index += 1;
        }
    }

    vm.load_slots(slots);

    return vm;
}

int main() {
    VendingMachine vm = initialize_vending_machine();

    vm.display_slots();

    int amount = 300;
    vm.insert_money(amount);

    std::cout
        << "Inserted $"
        << std::fixed << std::setprecision(2)
        << amount / 100.0
        << "."
        << std::endl;

    auto result = vm.purchase_item("A1");

    std::optional<Item> item = result.first;
    int change = result.second;

    if (item.has_value()) {
        std::cout
            << "Purchased " << item->name
            << ", received $"
            << std::fixed << std::setprecision(2)
            << change / 100.0
            << " as change."
            << std::endl;
    } 
    else {
        std::cout
            << "Purchase failed, received $"
            << std::fixed << std::setprecision(2)
            << change / 100.0
            << " as change."
            << std::endl;
    }

    assert(item.has_value());
    assert(item->name == "coke");
    assert(change == 150);
    assert(vm.get_slot_quantity("A1") == 4);
    assert(vm.get_inserted_balance() == 0);

    return 0;
}
